src/components/views/elements/Reply.js

```diff
@@ -0,0 +1,214 @@
+import SettingsStore from "../../../settings/SettingsStore";
+
+// This component does no cycle detection, simply because the only way to make such a cycle
+// would be to craft event_id's, using a custom homeserver; even then the impact would be low
+// as each event being loaded (after the first) is triggered by an explicit user action.
+export default class Reply extends React.Component {
+    static propTypes = {
+        // The parent event
+        parentEv: PropTypes.instanceOf(MatrixEvent),
+
+        onWidgetLoad: PropTypes.func.isRequired,
+    };
+
+    constructor(props, context) {
+        super(props, context);
+
+        this.state = {
+            // The loaded events to be rendered as linear-replies
+            events: [],
+
+            // The latest loaded event which has not yet been shown
+            loadedEv: null,
+            // Whether the component is still loading more events
+            loading: true,
+
+            // Whether as error was encountered fetching a replied to event.
+            err: false,
+        };
+
+        this.onQuoteClick = this.onQuoteClick.bind(this);
+
+        this.unmounted = false;
+    }
+
+    componentWillMount() {
+        this.room = this.getRoom(this.props.parentEv.getRoomId());
+        this.initialize();
+    }
+
+    componentWillUnmount() {
+        this.unmounted = true;
+    }
+
+    async initialize() {
+        const {parentEv} = this.props;
+        const inReplyTo = Reply.getInReplyTo(parentEv);
+
+        const ev = await this.getEvent(this.room, inReplyTo['event_id']);
+        if (this.unmounted) return;
+
+        if (ev) {
+            this.setState({
+                events: [ev],
+            }, this.loadNextEvent);
+        } else {
+            this.setState({err: true});
+        }
+    }
+
+    async loadNextEvent() {
+        const ev = this.state.events[0];
+        const inReplyTo = Reply.getInReplyTo(ev);
+
+        if (!inReplyTo) {
+            if (this.unmounted) return;
+            this.setState({
+                loading: false,
+            }, this.props.onWidgetLoad);
+            return;
+        }
+
+        const loadedEv = await this.getEvent(this.room, inReplyTo['event_id']);
+        if (this.unmounted) return;
+
+        if (loadedEv) {
+            this.setState({loadedEv}, this.props.onWidgetLoad);
+        } else {
+            this.setState({err: true}, this.props.onWidgetLoad);
+        }
+    }
+
+    getRoom(id) {
+        const cli = MatrixClientPeg.get();
+        if (id[0] === '!') return cli.getRoom(id);
+
+        return cli.getRooms().find((r) => {
+            return r.getAliases().includes(id);
+        });
+    }
+
+    async getEvent(room, eventId) {
+        const event = room.findEventById(eventId);
+        if (event) return event;
+
+        try {
+            await MatrixClientPeg.get().getEventTimeline(room.getUnfilteredTimelineSet(), eventId);
+        } catch (e) {
+            return;
```

Comment:
Our general rule of thumb is to prefer an explicit `return null` (or `undefined`).